src/ps/api/client/mod.rs

```diff
@@ -547,18 +554,37 @@ impl Pennsieve {
         &self,
         api_key: S,
         api_secret: S,
-    ) -> Future<response::ApiSession> {
-        let payload = request::ApiLogin::new(api_key.into(), api_secret.into());
-        let this = self.clone();
-        into_future_trait(
-            post!(self, "/account/api/session", params!(), &payload).and_then(
-                move |login_response: response::ApiSession| {
-                    this.inner.lock().unwrap().session_token =
-                        Some(login_response.session_token().clone());
-                    Ok(login_response)
-                },
-            ),
-        )
+        app_client_id: S,
+    ) -> Future<InitiateAuthResponse> {
+        let cognito = rusoto_cognito_idp::CognitoIdentityProviderClient::new(
+            rusoto_core::region::Region::UsEast1,
+        );
+
+        let mut auth_parameters = HashMap::<String, String>::new();
+        auth_parameters.insert("USERNAME".to_string(), api_key.into());
+        auth_parameters.insert("PASSWORD".to_string(), api_secret.into());
+
+        let request = InitiateAuthRequest {
+            analytics_metadata: None,
+            auth_flow: "USER_PASSWORD_AUTH".to_string(),
+            auth_parameters: Some(auth_parameters),
+            client_id: app_client_id.into(),
+            client_metadata: None,
+            user_context_data: None,
+        };
+
+        let mut runtime = tokio::runtime::current_thread::Runtime::new().unwrap();
+        let response = runtime.block_on(cognito.initiate_auth(request)).unwrap();
```

Comment:
If this method is returning a `Future`, why is it necessary to block in the `Future` body? Why not await the result of the `initiate_auth`(same as `and_then`), and then set the `session_token` to that of the access_token value? I very well could be mistaken since I haven't touched Rust in a while!

I get it if it's needed to get things to work (a common Rust problem!) If so, emitting a compiler warning and including a comment would suffice to warn future users that the `Future` returned by `login` will block the current thread, as it is unexpected behavior.

Also, I know the version of `tokio` that's being used is ancient at this point, but it's not necessarily safe to assume that the executor driving the chain of futures is the same implementation of `current_thread`. You can encounter race conditions when mixing executors.

Reply:
Yeah it was sort of a "get it working" hack. The `0.43` version of `initiate_auth` function returns this `Pin<Box<dyn Future<Output = Result<InitiateAuthResponse, RusotoError<InitiateAuthError>>> + Send + 'async_trait>>`. I'm not comfortable with working with Futures was having trouble unwrapping the types.

I'll definitely add a warning and I'll look into the point your raise about the executor. Thanks for the comprehensive review. 🙏